### ISSUE TITLE
Suggestion: Add support for Path objects from pathlib

### ISSUE THREAD

**Issue by Imanflow** [OSGeo/gdal#1613]
## Expected behavior and actual behavior.

When trying to open a file that I have as a Path objects, it fails.

And I have not found here anything related, so I don't thing anyone is planning to do it. 

## Steps to reproduce the problem.

from pathlib import Path
path = Path('/home/path/to/file/rasterfile.tif')
gdal.Open(path)

It returns:

RuntimeError: not a string

## Operating system

 Ubuntu 18.04 64 bit

## GDAL version and provenance

 the 2.4.1 version from ubuntugis-unstable PPA

 ## Comments:

As my understanding goes, this should be quite tribial change:

If input is Path object, convert to string, if it's string, continue, else, raise an Error.

So I think I could manage to do this.

**Comment by dericke** [comment]
> As my understanding goes, this should be quite tribial change:
>
>If input is Path object, convert to string, if it's string, continue, else, raise an Error.

It's even simpler than that. Just pass the input through the `str()` function. If it's already a str type, nothing changes, no exception is raised. If it's a Path, it'll become a str. No conditional needed.

**Comment by Imanflow** [comment]
Some time has passed... not sure if i missed the answer. Is this still an issue?

**Comment by RutgerK** [comment]
This would be a nice addition indeed. And since since GDAL 3.3 only supports Python 3.6 or higher, there is no longer a dependency issue with older Python versions. The `pathlib` module was introduced in Python 3.4, but in order to add support for this, you basically only need `os.fspath` which was introduced in Python 3.6.

Instead of using `str`, it's probably better to use `os.fspath`. Since you never know what the `__str__` method of a random object returns, GDAL would try to open it. 

I can see people creating a class where the string representation for example returns only a valid url. If you (accidentally) open that with GDAL it will try to open the url, and then probably fail if it's not a valid dataset. But catching it that late might be undesirable, since connections etc have already been set up.

Users who do want GDAL to be able to open their custom class can explicitly implement the `__fspath__` method.

```python
# Signature: os.fspath(path)
# Docstring:
# Return the file system path representation of the object.

# If the object is str or bytes, then allow it to pass through as-is. If the
# object defines __fspath__(), then return the result of that method. All other
# types raise a TypeError.
```

Example:

```python
class SomeClass:
    def __fspath__(self):
        return '\\some_folder\some_file.tif'
    
obj = SomeClass()

gdal.Open(os.fspath(obj))

# <osgeo.gdal.Dataset; proxy of <Swig Object of type 'GDALDatasetShadow *' at 0x0000019910476E70> >
```

**Comment by rouault** [comment]
CC @idanmiara

**Comment by idanmiara** [comment]
Great idea! I'm trying to implement it.

**Comment by dbaston** [comment]
I think this was largely been addressed in 3.8 (#8286, #8294, and others)
